### ISSUE TITLE
Relative images are relative to working directory, not file

### ISSUE THREAD

**Issue by Porges** [jgm/pandoc#3752]
I'm using 1.19.2.1.

I have a setup like:

* src
  * 001
    * 001.md
    * 001.jpg
  * 002
    * 002.md
    * 002.jpg

In the .md files I'm trying to include the images via relative links like `![](001.jpg)`, however when I do this and build from the top level (passing all the .md files as arguments), Pandoc cannot find the images. Instead I must supply the path as relative to the working directory (so `src/001/001.jpg`), which is a bit clunky.

So: a) is there any way to get my desired behaviour, and b) is the current behaviour intended? I would have expected paths to be relative to the file that they appear in.

Thanks!

**Comment by jgm** [comment]
Yes, current behavior is intended.  Pandoc just acts on a stream of text that may come from files (possibly several files in different directories) or from stdin; it doesn't keep track of what directory the text came from.  Thus `pandoc foo/bar.txt` is equivalent to `cat foo/bar.txt | pandoc`.

See #852, which added a `--resource-path` command line option.  This should help in your case, though it's not released. (You could try compiling from source or using pandoc-nightly, but be aware there are significant changes in 2.0.)  You may have problems with this, though, if you have files with the same name in different directories.

**Comment by jgm** [comment]
I can think of a somewhat complex way in which this might be improved.  Not sure it's worth it, though.  Instead of having the reader take a Text as argument, we could have it take something like a list of pairs of filenames and Texts:
```
data Source = Stdin | File FilePath | Url String
newtype Sources = Sources{ unSources :: [(Source, Text)] }
readMarkdown :: PandocMonad m => ReaderOptions -> Sources -> m Pandoc
```
Most of the readers use parsec parsers; we could define a custom Stream instance for `Sources` (or a newtype wrapper thereof) by defining
```
uncons :: s -> m (Maybe (t, s))
```
We could store the name of the current source file in the "common state" of the pandoc monad. The readers could then check for the image file, first in the local directory and then in the working directory or resource path, and adjust the path accordingly.  (Note: we don't normally do anything like this until the writers.)

This would help with your use case, at the expense of making the pandoc API considerably more complicated.  Not sure it's worth it.

@mb21 @jkr I'd be curious if you have any thoughts about this.  @jkr, can `--file-scope` help with this?  Perhaps when `--file-scope` is used we could automatically add the input file's path to resource path.  But this wouldn't really help, since file-scope only affects parsing, and currently we don't load resources until the writing phase.

**Comment by mb21** [comment]
That's a hard one. I've run into this (kind of unexpected behaviour) myself. Then again, it's really nice to have pandoc behave consistently when input is piped to in and when read from a file. With that in mind, I don't think making those intrusive and complicating changes is worth it.

**Comment by Porges** [comment]
The unfortunate thing is that this is inconsistent with the way GitHub processes Markdown. So when I make my paths relative to the working directory the images render as broken in the online view.

Maybe I'll just write a script to pre-process all my files into another directory before running pandoc...

**Comment by mb21** [comment]
another thought: we could abstract the file handling from the readers, so they would only get a mediabag or similar interface of files to query which could be instantiated with either files from the working directory or current source file directory – depending on a command line setting for example.

**Comment by jgm** [comment]
@mb21 Any change that would search for images in the working directory of the source file (when multiple files are specified on the command line) would have to keep track of which source file includes the given image, so we'd need the more complex interface I sketched above.

I did have one thought for a more limited change:  perhaps we could automatically set the resource path for images to include the directory of the first file argument. When `--file-scope` is used, we could set this for *each* file argument.  I think that would work for this use case.

**Comment by Wolf-at-SO** [comment]
@mb21 isn't this helpful also for multi-target publishing scenarios (HTML vs. PDF) where you need images in different resolution, which could be accomplished by changing folders? I find using the `--default-image-extension=` somewhat cumbersome.

**Comment by Porges** [comment]
@jgm I did actually try `--file-scope` in the hope that it would work. The downside is the lack of being able to link across files in that case, but I might be able to live with that (as long as it doesn't break `pandoc-citeproc`).

**Comment by Prajjwal** [comment]
`--file-scope` does not alleviate the issue.

A possible workaround for now could be writing a wrapper script that cd's into the correct directory, generates an output file for each input, and later concatenates the result. This is cumbersome, and I'm not even sure if it can be done with `ePub` output.

Another equally cumbersome workaround could be to pipe the markdown to `sed` and replacing the relative URLs with absolute ones.

I suppose we shall have to wait for the `--resource-path` option to hit stable and hope it works.

**Comment by archonic** [comment]
Does anyone know if `--resource-path` has hit stable yet?

**Comment by snow212-cn** [comment]
Not yet, and never it will

**Comment by mb21** [comment]
Actually, `--resource-path` is already released. See the [MANUAL](http://pandoc.org/MANUAL.html) for usage.

**Comment by Porges** [comment]
I guess to emulate my desired behaviour I can provide all the directories as `--resource-path` arguments (as long as I don't duplicate any filenames...)

**Comment by agusmba** [comment]
> I guess to emulate my desired behaviour I can provide all the directories as `--resource-path` arguments (as long as I don't duplicate any filenames...)

@Porges, depending on your workflow, wrappers that automatically cd into different directories such as `pandocomatic` could be of use to you.

**Comment by Porges** [comment]
@agusmba I'm generating a single file output from all the input files, so I don't think that would work.

**Comment by jacebenson** [comment]
I came across this problem when exploring using Pandoc.  I ended up renaming each image to something custom like the endpoint or the title of the post hypen image name.  Then I ran this script to generate the command to run.  YMMV https://gist.github.com/jacebenson/f6eba3a293def19bf8184defbf274dcc

**Comment by Agarwal-Nikhil** [comment]
@tarleb by using --resource-path=.. Images defined using markdown syntax works but those with latex syntax don't work.

**Comment by qaisjp** [comment]
I've run into this issue as well. [My directory structure is like this](https://user-images.githubusercontent.com/923242/73508885-413f8b80-43d5-11ea-9871-deb0fbb99a9f.png). And I just define input files like `chapters/**/*.md` which is (probably) expanded by Bash to each markdown file deep inside the `chapters` folder.

A filter like this one, except for pandoc 2, would work nicely. Except we'd still need a way to determine what folder our markdown files are in. https://github.com/jgm/pandoc/issues/3342#issuecomment-271139307

**Comment by snow212-cn** [comment]
Pandoc is a terrible thing for converting markdown to epub. To realize this desire, I recommend gitbook-cli instead.

**Comment by mb21** [comment]
> @mb21 Any change that would search for images in the working directory of the source file (when multiple files are specified on the command line) would have to keep track of which source file includes the given image, so we'd need the more complex interface I sketched above.

I keep re-reading this sentence and not sure I get it. I mean, I get:

> would have to keep track of which source file includes the given image

but why does this imply that:

> we'd need the more complex interface I sketched above

---

Why do we have to call the reader with a list of inputs `[(Source, Text)]`? Couldn't we do this similarly to how [`--file-scope` is implemented](https://github.com/jgm/pandoc/blob/659ee981764b85fb46845c086e3b10f1fc57a712/src/Text/Pandoc/App.hs#L266)? Just call the reader multiple times in App.hs, once for each input file, but each time with the resource-path (in the PandocMonad env) set to the file's directory.


There seem to be four modes:


<table class="tg">
  <tr>
    <th class="tg-0pky"></th>
    <th class="tg-0pky">paths are relative to cwd (--resource-path=.)</th>
    <th class="tg-0pky">paths are relative to each input file</th>
  </tr>
  <tr>
    <td class="tg-0pky">share reader state across files</td>
    <td class="tg-0pky">pandoc</td>
    <td class="tg-0pky"></td>
  </tr>
  <tr>
    <td class="tg-0pky">reset reader state after call to each reader</td>
    <td class="tg-0pky">pandoc --file-scope</td>
    <td class="tg-0pky"></td>
  </tr>
</table>

I'm not sure what's the best name for the option that would enable the "paths are relative to each input file"-behaviour. Feels a bit like a static site generator works... you might also want an option to give pandoc an input directory instead of individual input files, then it would default to this new mode.

**Comment by qaisjp** [comment]
> you might also want an option to give pandoc an input directory instead of individual input files, then it would default to this new mode

I already _kind_ of do this but lean on the shell to expand globs: `pandoc chapters/**/*.md`

**Comment by jgm** [comment]
@mb21 if you do it the way we do with file-scope, calling the reader for each one, then things like reference links and footnotes won't work between files (as with file-scope).

**Comment by mb21** [comment]
> if you do it the way we do with file-scope, calling the reader for each one, then things like reference links and footnotes won't work between files (as with file-scope).

@jgm I meant not do it exactly the same way... but in a similar way: call the reader multiple times in App.hs, once for each input file, but share/pass the state along each time. I guess that would mean exporting a second function from each reader, e.g. `readMarkdown opts carryOverState text`, or somehow provide an ability in PandocMonad for a reader to store it's state for when it's called the next time...

**Comment by jgm** [comment]
It's tricky because state types are very heterogeneous.  I suppose one approach would be to put something in PandocMonad for the inputs: `[(FilePath, Text)]`.  The parsers could then be rewritten so that after parsing blocks, they move to the next item in this list, reset the source position, and add the text to the input stream, then parse again.  This would require a lot of rewriting of parsers (and maybe there'd be problems in some cases) but the reader types could at least stay stable.

Indeed, a lot of this logic is already included in our current facility for handling include files in latex, RST, and other formats.

**Comment by brainchild0** [comment]
It seems that the original request expresses a simple but common case of a single document file  referencing external resources.  While features such as  concatenation of inputs are  useful generally, the interest of user friendliest might indicate the inclusion of options to help manage such plain usages.

I wonder whether this problem largely can be  solved with a new option.  When the option is present, then the processing of the regular sequence of input files from positional arguments  would be suppressed. The  argument for the option is a file, which is taken as the sole input file. Resources then are resolved relative to this file (whether through internally setting the resource path, or some other means).

**Comment by jgm** [comment]
With the recent changes in reader types, the parsers now have access to the path of the file containing an image link.  So we could now support these relative paths:  e.g., if `foo/test.md` contains `![image](img.png)`, `img.png` could be sought in the `foo` subdirectory.

Since this is a common pain point (note the 12 +1s on the original post above), I'd like to support this somehow, but backwards compatibility is also a concern.

The simplest approach would be to change the behavior so that pandoc defaults to looking in the directory containing the file with the image (I'll call this the "file directory"), rather than the working directory.  However, this could break some old workflows that assume the old behavior.  In addition, we'd have to figure out how it interacts with `--resource-path`. Currently, if you want to include the working directory in the resource path, you have to specify it explicitly.  If we wanted to do something similar with the file directory, we'd need a way to specify it in the resource-path:  perhaps with a special symbol like `@`.

Ideas welcome on how to handle this as cleanly and simply as possible.

**Comment by jgm** [comment]
@tarleb I wonder if you have thoughts on this?

**Comment by brainchild0** [comment]
> The simplest approach would be to change the behavior so that pandoc defaults to looking in the directory containing the file with the image (I'll call this the "file directory"), rather than the working directory. However, this could break some old workflows that assume the old behavior.

Any reason for not setting the new behavior to be switchable?




>  If we wanted to do something similar with the file directory, we'd need a way to specify it in the resource-path: perhaps with a special symbol like `@`.

If you are suggesting that `@foo` means _do something special with file `foo`_ rather than _do something normal with file `@foo`_, then I would suggest against it. Maybe I misunderstand the suggestion.

**Comment by jgm** [comment]
The idea would be that you could specify `--resource-path="@:.:images"`, and this would mean: first look in the directory of the containing file (@), then in the current working directory (.), then in `images`.

**Comment by brainchild0** [comment]
That's a bit safer, then, but the problem of path separators has been discussed, as it may work in system paths but less so for user data files. Repeating the same switch in sequence is probably a better solution.

By the way, "the directory containing the file" is a moving target, right? Since the input may include multiple files, the substitution is different for each one.

**Comment by jgm** [comment]
> By the way, "the directory containing the file" is a moving target, right? Since the input may include multiple files, the substitution is different for each one.

Yes, exactly.  And this was something we couldn't handle until some very recent changes which allow us to track positions better.

**Comment by brainchild0** [comment]
Yes, I understand about the new code structure, but the proposed interface semantics may be confusing, because they mix values of static versus dynamic resolution. If a document holds an intrinsic relationship to resources surrounding it, in a shared file tree, then the need for a list of static paths applied to each document consistently but separately begins to seem cumbersome.

**Comment by tarleb** [comment]
> @tarleb I wonder if you have thoughts on this?

I haven't formed a full opinion on the matter yet, but I very much like the idea of using `--resource-path` to solve this!

**Comment by brainchild0** [comment]
The starting point for discussion may be whether *resource path* continues in the new context to serve a function as useful as it had done originally, or whether in this case resolving resources relative to individual files weakens the case for *resource path* to such a degree that it might become considered vestigial.

**Comment by jgm** [comment]
I do agree that if we default to seeking images relative to the files containing the image links, `--resource-path` won't be as important.  But keep in mind that different formats work differently in this respect: for example, LaTeX includes are always processed relative to the top level (not the file containing the `include` command), and LaTeX images are sought in a search path (for this we use the resource path).  So we do need to keep `--resource-path` around -- also for backwards compatibility.

**Comment by brainchild0** [comment]
Yes, I understand, and agree. My suggestion is not directly to eliminate or even to deprecate *resource path*, as much as make a distinction between the strategy of adapting it to a new context versus discovering which features   directly play the most useful role  in that context.

Backward compatibility is a separate issue. Of course there should be a way to invoke the old behavior, with the old constructs, regardless of which new behavior is now available.

If for the sake of argument, you were starting initially with the   behavior of resolving resources relative to the file that references them, what interface would you then choose? The answer to this question has value in answering the question you have asked, even while legacy considerations inevitably must be integrated into the ultimate solution.

**Comment by jgm** [comment]
I'm currently tempted to do the following:

- For formats where this is natural and expected (Markdown, RST, probably Org, others?), *always* seek images relative to the directory of the file containing the image link.  If the image is not found there, use the resource path (unchanged from before).

- For formats like LaTeX where images are never sought relative to the file containing the image link, just use the resource path as before.

This could in theory cause backwards-compatibility problems, but only in very rare and strange circumstances (where someone has foo/bar.md referring to an image "img.png", and also has foo/img.png, but means for ./img.png to be selected instead).  It would automatically "do the right thing" in most cases, and it wouldn't require any changes to resource path's current behavior.  (Though down the line, with this change, we might want to change the default resource path to empty instead of `["."]`).

This feels like the right change to me.  However, doing this would require significant architectural changes. Currently, we only use `fetchItem` to seek resources when they are needed in the writers (or in things like SelfContained).  But the writers just have access to a parsed AST, which could have been produced from any source format (or constructed programatically).  So the writers can't behave differently depending on the source format, and they can't know which directory a given image link was found in.

One solution to this would be to use `fetchItem` in the readers, populating the "MediaBag" with all the resources that might be needed by the writer.  This would have a significant performance impact (since currently many conversions never need to load the actual files, e.g. in markdown -> html you only need to preserve the link).  But perhaps that could be mitigated by exploiting laziness.  I think I discussed this idea long ago with @jkr when we were implementing the MediaBag mechanisim, but I don't recall what we concluded, if anything.

Another solution would involve annotating the AST with information about the directory in which to seek resources. This seems ugly to me.  But it could be done without a breaking AST change, by adding special attributes to images to hold the paths.  We'd need to modify writers so that they don't render these special attributes.

**Comment by cagix** [comment]
Would it be an idea to look for the images in the resource path first if `--resource-path` is set, and as fallback or else (if not set) relative to the Markdown file? This way we could achieve both and would even be backwards compatible ...

**Comment by jgm** [comment]
I've pushed an `images2` branch that implements my current best approach to this, in case anyone wants to try it out.

**Comment by jgm** [comment]
@tarleb  (and others) I'd be curious what you think of this branch.  Here's the main idea:

- If an image `![alt](foo.jpg)` appears in a file `subdir/doc.md` (where `subdir` is relative to the working directory), it is parsed as having the path `subdir/foo.jpg`; in addition, an attribute `basename` is added to store the original `foo.jpg`.
- When seeking the content of the image in a writer, we first try the path (which here would be `subdir/foo.jpg`); if not found, we fall back to searching the resource path with `basename` (if present).  (Actually, in the current implementation we search the resource path with the full path first, then fall back. Not sure if that matters.)
- Rather than putting this image search logic in `fetchItem` (which would require giving it an additional parameter for `basename` and modifying all the writers which call it -- in some cases this is difficult because attributes are no longer easily accessible at the call site), I modify `fillMediaBag` so that it uses this logic.  We then ensure that `fillMediaBag` is called at the beginning of each of the writers that use `fetchItem` to get image data.
- One side effect of this approach is that we get hashes for attachments instead of the original filenames the ipynb writer -- since `fillMediaBag` uses hashes for the media keys.  Not sure that will matter.

The overall effect of these changes is that relative image links work more as expected in markdown. For example, if you have
```
chapter1/content.md
chapter1/fig1.png
chapter2/content.md
chapter2/fig1.png
```
where `chapter1/content.md` contains `![my figure](fig1.png)` and `chapter2/content.md` contains `![my figure](fig1.png)`,
```
pandoc chapter*/*.md -o doc.docx
```
will find the images in the expected locations.

**Comment by tarleb** [comment]
I'm all for it! The behavior seems intuitive, and the additional code complexity is well justified. It's quite beautiful and far less complex than I expected.

**Comment by jgm** [comment]
Note that anyone who has been using `-f markdown -t markdown foo/bar.md` as a prettifier will have unexpected results, as a directory will be added to each image link on each invocation. Is this a problem?

**Comment by jgm** [comment]
Here is one alternative solution I considered:  instead of modifying the image destination and adding `basename` to track the original, we could keep the destination the same and add `directory` to track the directory containing it.  `fillMediaBag` could be modified to handle this alternative structure; in a way it might even be cleaner.  (And it wouldn't be so problematic with the `-f markdown -t markdown` case above.)

I think the reason I went for the other approach was to get better results for formats that don't need to load image data, e.g. plain HTML.  If you do `pandoc foo/bar.md -o bar.html`, and the image lives in `foo/`, then you'll need the img src to start with `foo/`.

**Comment by jgm** [comment]
If we took this alternative approach (which is more conservative), we could consider adding a command-line option `--rebase-links` (or something) that would rewrite Image and Link destinations in the way the images2 branch currently does.  That could be put off for the future, though.

**Comment by jgm** [comment]
An even more conservative approach:

- add a command-line option to rewrite the image paths the way `image2` branch currently does...
- ...but possibly without adding `basename` (which was needed mainly for backwards compatibility)
- don't make any change in the  behavior of `fillMediaBag`, but *do* use `fillMediaBag` before the writers that need to load image content, as the `image2` branch does.

With these changes, people who don't use the option would not see changes (except for hashes used as path names for ipynb attachments).  People who want to convert nested file trees with relative image links would just have to know to use the option.

Tempted to try this for 2.14...

**Comment by jgm** [comment]
Maybe the option should have an argument.
```
pandoc --rebase-relative-paths="." subdir/test.md
```
would rewrite `![image](img.jpg)` in `subdir/test.md` as if it were `![image](subdir/img.jpg)`.

```
pandoc --rebase-relative-paths="subdir" subdir/test.md subdir/subsubdir/extra.md
```
would rewrite `![image](img.jpg)` in `subdir/test.md` as `![image](img.jpg)` (i.e. unchanged) and `![image](img.jpg)` in `subdir/subsubdir/extra.md` as `![image](subsubdir/img.jpg)`.

**Comment by jgm** [comment]
Presumably this option should affect both Link and Image?  Is this a good name for it?

**Comment by brainchild0** [comment]
Quite a few ideas and details were presented, and parsing all the permutations is difficult, by but my favorite overall is the earlier one, as follows:

> For formats where this is natural and expected... *always* seek images relative to the directory of the file containing the image  link.  [Otherwise], just use the resource path as before.

I think that targeting a new behavior that is, as you put it, "natural and expected", for a particular format, is an appropriate choice, even if the cost is not small.

Backwards compatibility is no small concern, but the reasons given for breaking it in this case are sound. 

To the details of this dilemma, I would add the following few small points:

- A warning message is probably favorable, for invocations that are affected by the new behavior. If possible, you might consider limiting the warning to cases where the file exists in both locations that would be selected in the old versus new behavior. Otherwise, you might just generalize it to cases where the affected formats are used. The warning message may indicate the new behavior, and may itself be eventually phased out of the application.
- You might include an option to preserve the earlier behavior. I know that the growing list of options has been a concern, but you could keep the extra option as a hidden feature, mentioned in the above warning, which also could be removed eventually.

You mentioned a few other concerns about compatibility, but I will leave my comments on the question for now, in the interest of brevity.

**Comment by brainchild0** [comment]
> (Actually, in the current implementation we search the resource path  with the full path first, then fall back. Not sure if that matters.)

I'm not sure I follow, but I would always expect the test path for the target file to be the concatenation of the resource path item and the complete path given in the document. If the idea is that *a/b + c/d => a/b/d*, then I would say it feels to me as not the desirable behavior.

**Comment by brainchild0** [comment]
I lack a background on the architectural matters, and am unable to read Haskell, but generally I think now as in the past an issue has been a legacy architecture that is deeply tied to specific assumptions about invocation environment. 

With respect to questions such as resolving resources external to the document, and other matters relating to system and input/output, I suppose an ideal design is one in which the format details represented by the readers and writers are abstracted separately from details like resource resolution, the latter functionality being provided by callback abstractions given by the context that invokes the reader or writer. 

As such, when invoking a reader a reader, for example, the application (or any other invoking context) would provide the input (text or bytestring data) as well as the callbacks that represent the functionality for resolving resources, as appropriate to that context.   Under such a model, changes in the invocation context, or broadened support for various contexts, requires no changes to the readers, because of an existing separation of concerns.

I imagine thses remarks may seem off topic, but in case they are helpful or relevant, I include them.

**Comment by jgm** [comment]
Just to clarify, there are three basic options on the table here.

Option 1 (implemented in the current `images2` branch) actually rewrites relative paths for images, so that an image `![image](a.jpg)` in `b/x.md` will be parsed as if it were (in current pandoc) `![image](b/a.jpg)`.  This will have implications both for loading binary resources (in files that embed binary images, such as docx) and for image links (in files that just embed links to images, such as HTML).

Option 2 would be to leave relative image paths as they are, not rewriting them, but to take the directory within which the image link is found into account in loading binary resources.  With this option, `![image](a.jpg)` in `b/x.md` would produce a link to `a.jpg` in HTML, but in creating a docx we'd look for image data in `b/a.jpg`.

Option 3 would be creating a command-line option that causes Option 1 behavior; if the option is not supplied, then pandoc acts as it does now.

Option 1 has the advantage of "doing the right thing" automatically in many cases.  But perhaps not in all.  If you do `pandoc subdir/a.md -o subdir/a.html`, for example, you're not going to want this image path rewriting behavior.

Option 2 is more conservative and will probably be harmless in most cases -- though one can think of scenarios where it will lead to unexpected results.  But there are cases where the link-rewriting behavior is what you'd want (e.g. in the HTML example motivating Option 1 above), and it will not give you that.  It will only help with output in binary formats that include image data.

Option 3 is the most conservative, since it requires an explicit decision to depart from current behavior.  But it *allows* you to get the Option 1 link-rewriting behavior, and in that respect it gives you more power than Option 2.

**Comment by brainchild0** [comment]
The way I evaluate the overall issue is as two broad questions, one being how to find resources referenced by an input document, and the other being how to generate  an output document referencing external resources.

 Most generally, one way to generate an HTML result is to build a new directory tree independent of the source, similar to the operation of browsers saving a web page. This behavior is different from current operation of Pandoc, but is worth considering, because it points to the general independence of how resources are gathered versus what eventually gets created.

More immediately, the difference between options 1 and 2  seems to be one ideally handled  by the (HTML) writer, since in both cases the target location for the resources referenced in the input document is resolved the same.

Perhaps a comprehensive solution to the broad problem involves independent management of resource locations for both writers versus readers.

**Comment by tarleb** [comment]
I like all options, with option 3 being my favorite.

**Comment by jgm** [comment]
On option 3, any opinions about the option name?

**Comment by jgm** [comment]
Here's what I've come up with for documentation for the proposed feature:

`--rebase-relative-paths=`*PATH*

Rewrite relative paths for Link and Image elements, depending
on the path of the file containing the link or image link.
For each link or image, pandoc will compute the directory of
the containing file, relative to the directory given as argument of
`--rebase-relative-paths`, and prepend the resulting directory
to the path.  The use of this option is best understood by
example. Suppose you have a a subdirectory for each chapter
of a book, `chap1`, `chap2`, `chap3`. Each contains a file `text.md`
and a number of images used in the chapter.  You would like
to have `![image](spider.jpg)` in `chap1/text.md` refer to
`chap1/spider.jpg` and `![image](spider.jpg)` in `chap2/text.md`
refer to `chap2/spider.jpg`.  To do this, use

    pandoc chap*/*.md --rebase-relative-paths=.

Without this option, you would have to use
`![image](chap1/spider.jpg)` in `chap1/text.md` and
`![image](chap2/spider.jpg)` in `chap2/text.md`.  Links with
relative paths will be rewritten in the same way as images.
When image paths are rewritten, the original path will be
added in an attribute `orig`.  This will be used as a
fallback when pandoc seeks image data on the resource path
(see `--resource-path`).  *This option currently only affects
Markdown input.*

**Comment by jgm** [comment]
As for implementation, I'm thinking this would involve a new reader option and changes to the Markdown reader. It would also be possible to implement this as an AST transform after the reader, but this would  still require changes to the reader, which would have to insert the needed information about the source location of the image or link elements.

**Comment by qaisjp** [comment]
Btw I think there's an ambiguity in `--rebase-relative-paths=.` — I believe it's easy to confuse it with `--rebase-relative-paths=$(pwd)`.

**Comment by brainchild0** [comment]
> On option 3, any opinions about the option name?

`--resource-path-target`, `--resource-target-path`, `--resource-base`.

I would suggest `--resource-path-base` or `--resource-base-path`, but I suppose they are too easy to confuse with the existing option.

**Comment by jgm** [comment]
> `--rebase-relative-paths=.` — I believe it's easy to confuse it with `--rebase-relative-paths=$(pwd)`

I'm not quite sure what you have in mind here. `.` and `$(pwd)` refer to the same directory, after all.  But there is a question whether these two things should do the same thing, or something different.  E.g. should a reference to `img.jpg` occurring in `a/x.md` be rewritten to `a/img.jpg` with the former and `/home/user/docs/a/img.jpg` (or whatever) with the latter?

I'm somewhat tempted to drop the argument and just always make the rewriting relative to the working directory. I can think of a few cases where the argument could be helpful, but in all those cases you could just change directories before calling pandoc to deal with the issue.

`--resource-path` or anything like that should be avoided. We have `--resource-path` already, and that refers to something different.

**Comment by qaisjp** [comment]
Oh, I thought `=.` was for the behaviour of "reading from the directory of the markdown file", not for "reading from the cwd of this invocation"

**Comment by jgm** [comment]
I've simplified this a bit, making the argument optional and removing the special attribute to facilitate use with `--resource-path`:

`--rebase-relative-paths=`[`=`*PATH*]

Rewrite relative paths for Link and Image elements, depending
on the path of the file containing the link or image link.
For each link or image, pandoc will compute the directory of
the containing file, relative to the working directory
(or, optionally, the directory given as argument of
`--rebase-relative-paths`), and prepend the resulting directory
to the path.

The use of this option is best understood by example.
Suppose you have a a subdirectory for each chapter of a
book, `chap1`, `chap2`, `chap3`. Each contains a file
`text.md` and a number of images used in the chapter.  You
would like to have `![image](spider.jpg)` in `chap1/text.md`
refer to `chap1/spider.jpg` and `![image](spider.jpg)` in
`chap2/text.md` refer to `chap2/spider.jpg`.  To do this,
use

    pandoc chap*/*.md --rebase-relative-paths

Without this option, you would have to use
`![image](chap1/spider.jpg)` in `chap1/text.md` and
`![image](chap2/spider.jpg)` in `chap2/text.md`.  Links with
relative paths will be rewritten in the same way as images.
*This option currently only affects Markdown input.*

`

**Comment by jgm** [comment]
Further simplification, removing the optional argument. (If there is a demand for this, we can always add it later without compromising backwards compatibility.)

`--rebase-relative-paths`

Rewrite relative paths for Link and Image elements, depending
on the path of the file containing the link or image link.
For each link or image, pandoc will compute the directory of
the containing file, relative to the working directory, and
prepend the resulting path to the link or image path.

The use of this option is best understood by example.
Suppose you have a a subdirectory for each chapter of a
book, `chap1`, `chap2`, `chap3`. Each contains a file
`text.md` and a number of images used in the chapter.  You
would like to have `![image](spider.jpg)` in `chap1/text.md`
refer to `chap1/spider.jpg` and `![image](spider.jpg)` in
`chap2/text.md` refer to `chap2/spider.jpg`.  To do this,
use

    pandoc chap*/*.md --rebase-relative-paths

Without this option, you would have to use
`![image](chap1/spider.jpg)` in `chap1/text.md` and
`![image](chap2/spider.jpg)` in `chap2/text.md`.  Links with
relative paths will be rewritten in the same way as images.
*This option currently only affects Markdown input.*

**Comment by jgm** [comment]
Implementing this would require:

- [ ] update manual
- [ ] add tests
- [ ] add reader option, readerRebaseRelativePaths
- [ ] modify markdown reader to be sensitive to this

**Comment by dstadelm** [comment]
I'm having the same issue with relative paths for ReST. The ReST documentation at https://docutils.sourceforge.io/docs/ref/rst/directives.html#including-an-external-document-fragment states 
>    The "include" directive reads a text file. The directive argument is the path to the file to be included, relative to the document containing the directive. Unless the options literal, code, or parser are given, the file is parsed in the current document's context at the point of the directive.

Therefore for a directory structure as follows
![image](https://user-images.githubusercontent.com/3010527/119484794-a94a7b80-bd56-11eb-93ae-c7f42abb0d15.png)

with the following contents:

*foo.rst*
> .. include:: bar/bar.rst

*bar.rst*
> .. include:: here.rst
> .. include:: there/file.rst

This should work if I'm reading the specification of reST correctly. I'm not sure about the comment of @jgm https://github.com/jgm/pandoc/issues/3752#issuecomment-309998353 , even if the main file passed could be from stdin such as using cat, the include statements still have an actual file location, and as the only file initially given, is the main file, the only assumption that would have to be made would be that about the location of the main file. In which case I would assume the working directory of pandoc.

However I've seen that docutils which defines this behaviour also doesn't comply with their own specification 😥

It would make life so much easier, as one could write modules containing documentations and move them around freely without worrying about the context they are living in.

**Comment by qaisjp** [comment]
One last piece of feedback from me, as a person who encountered this issue in the wild, I would also like to propose this:

**Hint for ambiguities**

Flags: nothing special
Where: `chap1/text.md` refers to `spider.jpg`
Files present: `/spider.jpg` and `chap1/spider.jpg`
Outcome: print a warning like 

```
[W] Reference "spider.jpg" in `chap1/text.md` is ambiguous. It currently refers to "/spider.jpg" but could also refer "chap1/spider.jpg". To pick the latter, provide `--rebase-relative-paths`. To suppress this warning, provide `--no-rebase-relative-paths`.
```


**Hint for broken references**

Flags: nothing special
Where: `chap1/text.md` refers to `spider.jpg`
Files present: `chap1/spider.jpg` but not `/spider.jpg`
Outcome: print a warning like 

```
[W] Reference "spider.jpg" in `chap1/text.md` could not be found at "/spider.jpg". If you meant to refer to "chap1/spider.jpg", provide `--rebase-relative-paths`.
```

**Ignore ambiguities if a specific flag is specified**

Flags: `--no-rebase-relative-paths` or `--rebase-relative-paths`
Where: `chap1/text.md` refers to `spider.jpg`
Files present: either of the above scenarios
Outcome: no warnings

----

The pattern of prefixing with `--no-` is derived from Git. It's been [over a year](https://github.com/jgm/pandoc/issues/3752#issuecomment-581134178) since i've used pandoc so I have zero context on how warnings work and what the CLI parameter patterns are.

One shortcoming of this proposal is that it doesn't work great if you want to opt one Markdoc file into rebasing relative paths, but not another, but I don't envisage that being a popular usecase.

**Comment by jgm** [comment]
@dstadelm this issue with RST includes is already fixed (#6632) but the fix isn't yet in an official release.  I've done some experiments on image links in RST, and it doesn't look as if RST rebases image links in the way I've described above; so far, then, this would be a Markdown-only feature.

**Comment by jgm** [comment]
Here's another option that should be considered:

Option 4:  No extra option. Do the image resolution (and load binary data into the media bag) in the Markdown reader.  Always look for the image first relative to the file containing the image link, and then in the resource path.  Modify the image path to match the first matching image.  Emit an INFO message indicating which path has been matched.  Emit a WARNING message if nothing is found (this is done already now, but in the writers, and only for formats that include image data and not just links).

Potential advantages:

- This would just "do the right thing automatically" in the vast majority of cases.
- It would allow the `--resource-path` to affect not just output formats like docx, but formats like HTML that employ image links; the resource path could affect what path is used in the link.

Potential drawbacks:

- Some users may find it objectionable that their image links get rewritten based on what is actually on the file system.
- We'd be doing unnecessary IO when converting to formats that don't require image data (like HTML); this would affect performance considerably for image-heavy conversions
- People who have set the resource-path so that it doesn't include the working directory might have breakages, since images would be sought relative to the including file (which could be the working directory) before the resource path is consulted.

**Comment by jgm** [comment]
I've pushed a `rebase-relative-paths` branch that implements Option 3, if anyone wants to try it out.

**Comment by jgm** [comment]
My new thought is that it makes more sense for this to be an extension.
This idea is developed in the `rebase-relative-paths-extension` branch.

**Comment by jgm** [comment]
This is in master branch now.  Currently the extension only works for `markdown`.  I am working on making it work with `commonmark` and `gfm`, but this requires changes in commonmark-hs.

**Comment by Altair-Bueno** [comment]
# TLDR

```tree
Book
   ├── resources
   │   └── A1.png
   ├── Chapter-1
   │   └── ej1.md
   ...
```

```markdown
<!-- ej1.md -->
# Hello world

![Image](../A1.png)
```

Render every chapter with:

```shell
pandoc "--from=markdown+rebase_relative_paths" --output=Book.pdf Chapter**/*.md
```

# More info

Search for `Extension: rebase_relative_paths` on pandoc's manual
